contracts/compressor/IntentCompressor.sol

```diff
@@ -0,0 +1,336 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.10;
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import {IInbox} from "../interfaces/IInbox.sol";
+import {IIntentSource} from "../interfaces/IIntentSource.sol";
+import {Intent, Call, TokenAmount, Route, Reward} from "../types/Intent.sol";
+
+struct EncodedIntent {
+    uint8 sourceChainIndex;
+    uint8 destinationChainIndex;
+    // Reward token
+    uint8 rewardTokenIndex;
+    uint48 rewardAmount;
+    // Route token
+    uint8 routeTokenIndex;
+    uint48 routeAmount;
+    // Expiry duration
+    uint24 expiryDuration;
+    // Salt
+    bytes8 salt;
+}
+
+struct EncodedFulfillment {
+    uint8 sourceChainIndex;
+    uint8 destinationChainIndex;
+    // Reward token
+    uint8 routeTokenIndex;
+    uint48 routeAmount;
+    // Prove type (INSTANT = 0, BATCH = 1)
+    uint8 proveType;
+    // Claimant
+    address claimant;
+}
+
+contract IntentCompressor {
+    using SafeERC20 for IERC20;
+
+    /**
+     * @notice Thrown when the vault has insufficient token allowance for reward funding
+     * @param token The token address
+     * @param spender The spender address
+     * @param amount The amount of tokens required
+     */
+    error InsufficientTokenAllowance(
+        address token,
+        address spender,
+        uint256 amount
+    );
+
+    address public immutable PROVER;
+    IInbox public immutable INBOX;
+    IIntentSource public immutable INTENT_SOURCE;
+
+    constructor(address _intentSource, address _inbox, address _prover) {
+        INBOX = IInbox(_inbox);
+        INTENT_SOURCE = IIntentSource(_intentSource);
+        PROVER = _prover;
+    }
+
+    function fulfill(
+        bytes32 payload,
+        bytes32 rewardHash,
+        bytes32 routeSalt
+    ) external returns (bytes[] memory) {
```

Comment:
should maybe be payable? and add native transfer to the recipient to ancoded calls if `msg.value > 0` ?

Reply:
Yeah, great idea as well 👌🏻